{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Previous final database contains 0 rows and 3 columns\n",
      "\n",
      "The directory selected is: /home/laura/DEVELOP/VCF_DDTB/test\n",
      "The file is: 6-49128_final.bam.snp.final\n",
      "The sample name retrieved is: 6-49128\n",
      "\n",
      "The sample 6-49128 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 6-49128\n",
      "\n",
      "SAMPLE:\t6-49128\n",
      "TOTAL Variants:\t776\n",
      "Shared Variants:\t0\n",
      "New Variants:\t776\n",
      "\n",
      "The file is: 6-85169_final.bam.snp.final\n",
      "The sample name retrieved is: 6-85169\n",
      "\n",
      "The sample 6-85169 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 6-85169\n",
      "\n",
      "SAMPLE:\t6-85169\n",
      "TOTAL Variants:\t813\n",
      "Shared Variants:\t764\n",
      "New Variants:\t49\n",
      "\n",
      "The file is: 6-48233_final.bam.snp.final\n",
      "The sample name retrieved is: 6-48233\n",
      "\n",
      "The sample 6-48233 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 6-48233\n",
      "\n",
      "SAMPLE:\t6-48233\n",
      "TOTAL Variants:\t809\n",
      "Shared Variants:\t800\n",
      "New Variants:\t9\n",
      "\n",
      "The file is: 14N1_final.bam.snp.final\n",
      "The sample name retrieved is: 14N1\n",
      "\n",
      "The sample 14N1 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 14N1\n",
      "\n",
      "SAMPLE:\t14N1\n",
      "TOTAL Variants:\t881\n",
      "Shared Variants:\t476\n",
      "New Variants:\t405\n",
      "\n",
      "The file is: 6-37190_final.bam.snp.final\n",
      "The sample name retrieved is: 6-37190\n",
      "\n",
      "The sample 6-37190 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 6-37190\n",
      "\n",
      "SAMPLE:\t6-37190\n",
      "TOTAL Variants:\t824\n",
      "Shared Variants:\t820\n",
      "New Variants:\t4\n",
      "\n",
      "The file is: 14N13_final.bam.snp.final\n",
      "The sample name retrieved is: 14N13\n",
      "\n",
      "The sample 14N13 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 14N13\n",
      "\n",
      "SAMPLE:\t14N13\n",
      "TOTAL Variants:\t881\n",
      "Shared Variants:\t881\n",
      "New Variants:\t0\n",
      "\n",
      "The file is: 1B3_final.bam.snp.final\n",
      "The sample name retrieved is: 1B3\n",
      "\n",
      "The sample 1B3 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 1B3\n",
      "\n",
      "SAMPLE:\t1B3\n",
      "TOTAL Variants:\t889\n",
      "Shared Variants:\t780\n",
      "New Variants:\t109\n",
      "\n",
      "The file is: 6-25213_final.bam.snp.final\n",
      "The sample name retrieved is: 6-25213\n",
      "\n",
      "The sample 6-25213 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 6-25213\n",
      "\n",
      "SAMPLE:\t6-25213\n",
      "TOTAL Variants:\t798\n",
      "Shared Variants:\t793\n",
      "New Variants:\t5\n",
      "\n",
      "The file is: 2B1_final.bam.snp.final\n",
      "The sample name retrieved is: 2B1\n",
      "\n",
      "The sample 2B1 is NOT in final ddbb. Adding sample\n",
      "Checking positions(SNPs) in sample 2B1\n",
      "\n",
      "SAMPLE:\t2B1\n",
      "TOTAL Variants:\t890\n",
      "Shared Variants:\t888\n",
      "New Variants:\t2\n",
      "\n",
      "Final database now contains 1359 rows and 12 columns\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import argparse\n",
    "\n",
    "\n",
    "\n",
    " #TODO: \n",
    "    #- better implementation of sample name\n",
    "    #- exit if file don't exist\n",
    "    #- create a file with the date as backup\n",
    "    #- handle when sample exist\n",
    "\n",
    "directory = \"/home/laura/DEVELOP/VCF_DDTB/test\"\n",
    "cwd = os.getcwd()\n",
    "\n",
    "\n",
    "\n",
    "final_ddbb = pd.read_csv(\"DDBB_TB_FINAL.csv\", sep=',', header=0)\n",
    "print(\"Previous final database contains %s rows and %s columns\\n\" % final_ddbb.shape)\n",
    "\n",
    "print(\"The directory selected is: %s\" % directory)\n",
    "\n",
    "\n",
    "for filename in os.listdir(directory):\n",
    "    \n",
    "    print(\"The file is: %s\" % filename)\n",
    "    \n",
    "    positions_shared = []\n",
    "    positions_added = []\n",
    "    \n",
    "    \n",
    "    \n",
    "    #Manage sample name. Split by \"_\" and take the first word on the left\n",
    "    ###################\n",
    "    sample = filename.split(\"_\")[0]\n",
    "    \n",
    "    print(\"The sample name retrieved is: %s\\n\" % sample)\n",
    "    \n",
    "    \n",
    "    #Manage file[s]. Check if file exist and is greater than 0\n",
    "    ###############\n",
    "    file = os.path.join(directory, filename) #Whole file path\n",
    "    file_info = os.stat(file) #Retrieve the file info to check if has size > 0\n",
    "\n",
    "\n",
    "    if os.path.isfile(file) and file_info.st_size > 0:\n",
    "        \n",
    "        pass\n",
    "\n",
    "    else:\n",
    "        print(\"ERROR: Your file %s does not exist or is empty\" % filename)\n",
    "\n",
    "    #Import files in annotated vcf format\n",
    "    #####################################\n",
    "\n",
    "    new_sample = pd.read_csv(file, sep='\\t', skiprows=[0], header=None)\n",
    "\n",
    "    \n",
    "    #Handle each new_sample\n",
    "    #######################\n",
    "    \n",
    "    #print(\"This file contains %s SNPs and %s columns\" % new_sample.shape)\n",
    "    \n",
    "    \n",
    "    #Check if sample exist\n",
    "    ######################\n",
    "    \n",
    "    if sample not in final_ddbb.columns:\n",
    "        print(\"The sample %s is NOT in final ddbb. Adding sample\" % sample)\n",
    "        #print(\"Adding new sample %s to final ddbb\" % sample)\n",
    "        \n",
    "        #extrac the number of columns to insert a new one\n",
    "        new_colum_index = len(final_ddbb.columns)\n",
    "        #final_ddbb[sample] = sample #adds a new column but fills all blanks with the value sample\n",
    "        \n",
    "        #add a new column with defauls values = 0\n",
    "        final_ddbb.insert(new_colum_index, sample, 0)\n",
    "        \n",
    "        #Check if position exist\n",
    "        ########################\n",
    "        \n",
    "        print(\"Checking positions(SNPs) in sample %s\" % sample)\n",
    "        for position in new_sample.iloc[:,0].unique(): #extract first column in file\n",
    "\n",
    "            #print(type(position))\n",
    "            #print(final_ddbb[\"Position\"].values.astype(int))\n",
    "            \n",
    "            \n",
    "            if position not in final_ddbb[\"Position\"].values:\n",
    "               \n",
    "                positions_added.append(position)\n",
    "                \n",
    "                new_row = len(final_ddbb.index)\n",
    "                final_ddbb.loc[new_row,'Position'] = position\n",
    "                final_ddbb.loc[new_row,'Samples'] = sample\n",
    "                final_ddbb.loc[new_row,'N'] = int(1)\n",
    "                final_ddbb.loc[new_row,sample] = str(1)\n",
    "                \n",
    "            else:\n",
    "                \n",
    "                positions_shared.append(position)\n",
    "                \n",
    "                #Check whether the column matches the value and retrieve the first position [0]\n",
    "                #of the object index generated\n",
    "                index_position = final_ddbb.index[final_ddbb[\"Position\"] == position][0]\n",
    "                \n",
    "                                              \n",
    "                number_samples_with_position = final_ddbb.loc[index_position,'N']\n",
    "                names_samples_with_position = final_ddbb.loc[index_position,'Samples']\n",
    "                new_names_samples = names_samples_with_position + \",\" + sample\n",
    "                \n",
    "                #print(\"NEW NAME = %s\" % new_names_samples)\n",
    "                \n",
    "                final_ddbb.loc[index_position,'N'] = number_samples_with_position + 1 #add 1 to the number of samples\n",
    "                final_ddbb.loc[index_position,'Samples'] = new_names_samples\n",
    "                final_ddbb.loc[index_position,sample] = int(1) #Add \"1\" in cell with correct position vs sample\n",
    "    \n",
    "    else:\n",
    "        \n",
    "        print(\"This sample ALREADY exist, have a look\")\n",
    "              \n",
    "    #Create small report with basic count\n",
    "    #####################################\n",
    "    \n",
    "    #print(\"%s new positions from %s were added to final ddbb: %s\" % (len(positions_added), sample, positions_added))\n",
    "    #print(\"%s positions from %s were already in final ddbb: %s\" % (len(positions_shared), sample, positions_shared))\n",
    "    \n",
    "    print(\"\\nSAMPLE:\\t%s\\nTOTAL Variants:\\t%s\\nShared Variants:\\t%s\\nNew Variants:\\t%s\\n\" % (sample, len(new_sample.index), len(positions_shared), len(positions_added)))\n",
    "    \n",
    "#Finally NaN are replaced by 0, decimals are not displayed and posiitons are sorted\n",
    "#df.astype(int) AND set_option('precision', 0) doesn't work with string cells\n",
    "\n",
    "#final_ddbb = final_ddbb[\"Position\"].astype(int)\n",
    "pd.set_option('display.precision', 0)\n",
    "#pd.reset_option('^display.', silent=True) #Reset options in case I mess up\n",
    "\n",
    "#pd.set_option('precision', 5)\n",
    "#pd.options.display.float_format = '{:,2f}'.format\n",
    "\n",
    "final_ddbb = final_ddbb.fillna(0).sort_values(\"Position\")\n",
    "    \n",
    "print(\"Final database now contains %s rows and %s columns\" % final_ddbb.shape)\n",
    "#print(final_ddbb)\n",
    "\n",
    "final_ddbb.to_csv(\"final_aprin_2019.csv\", sep='\\t', index=False)\n",
    "\n",
    "#Create a Pandas Excel writer using XlsxWriter as the engine.\n",
    "#writer = pd.ExcelWriter(\"pandas_column_formats.xlsx\", engine='xlsxwriter')\n",
    "\n",
    "#Convert the dataframe to an XlsxWriter Excel object.\n",
    "#df.to_excel(writer, sheet_name='Sheet1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
